### ISSUE TITLE
Named export vs Default export

### ISSUE THREAD

**Issue by bidoubiwa** [meilisearch/instant-meilisearch#122]
# Named or Default

As of today, `instant-meilisearch` uses default export. This means that in order for a user to use instant-meilisearch in an ES environment he has to import it this way: 
```javascript
import instantMeiliSearch from 'instant-meilisearch'
```
The suggestion this issue considers is changing from this default export to a named export:
```javascript
import { instantMeiliSearch } from 'instant-meilisearch'
```

Should we change our default export to a named export? 
Here are a few arguments, if you have any other feel free to share it here/

## Pro's  of default search

- More common, or at least they used to be more common. There is no exact data on it. Although, I observe a change in this tendency. 
- Can be considered **Easier to use**, as only one export where you can change the name offers no complication to use it. `import whatIwant from 'instant-meilisearch'` instead of `import { instantMeiliSearch } from 'instant-meilisearch'`.
    Which can also be done with named exports `import { instantMeiliSearch as whatIWant } from 'instant-meilisearch'`, which if you use VS code should be auto-completed and thus present no difficulties to do.
- Changing it now would be breaking change (but the earlier, the better?) @bb 
- It seems that Typescript suggest using default exports in the [red flag category](https://www.typescriptlang.org/docs/handbook/modules.html#red-flags) of their documentation. But this is not really clear and based on [this stackOverFlow post](https://stackoverflow.com/questions/33305954/typescript-export-vs-default-export#:~:text=For%20example%2C%20if%20you%20default,in%20all%20the%20other%20files) where a user that prefers named exports mentions the redflag.
- No changes needed in the code as it is already a default export
- [vue-instantsearch](https://github.com/algolia/vue-instantsearch) and [algoliaSearch](https://github.com/algolia/instantsearch.js) uses default export **but not** [react-instantsearch](https://github.com/algolia/react-instantsearch)
    ```
    import InstantSearch from 'vue-instantsearch';
    import algoliasearch from 'algoliasearch/lite';
    import { InstantSearch, SearchBox, Hits } from 'react-instantsearch-dom';
    ```
## Con's of default search

The first 5 arguments comes from this [typescript book](https://github.com/basarat/typescript-book) that has 12k stars on github we learn that:  [default is bad](https://basarat.gitbook.io/typescript/main-1/defaultisbad). I would suggest reading the article as it showcases it better.

1. [Poor discoverability](https://basarat.gitbook.io/typescript/main-1/defaultisbad#poor-discoverability): intellisense works best with named exports. [For example this issue](https://stackoverflow.com/questions/54427609/vscode-intellisense-not-fully-working-with-es6-imports)
2. Because of `1.`, [lack of autocompletion](https://basarat.gitbook.io/typescript/main-1/defaultisbad#autocomplete)
3. [CommonJS interop](https://basarat.gitbook.io/typescript/main-1/defaultisbad#commonjs-interop): Node users have an horrible experience because of the way default exports are transpiled into cjs:  `const {default} = require('module/foo');`. This can be avoided thanks to bundlers ( like rollup ). So this is not a problem for us. meilisearch/meilisearch-js#661
4. [Avoiding typos](https://basarat.gitbook.io/typescript/main-1/defaultisbad#typo-protection): Only one way to import, avoiding having it imported differently in multiple files of a same project `import Foo from 'foo'`, `import foo from 'foo'`. Instead you have to import it with the right syntax `import { Foo } from 'foo'`
5. [Bad compatibility with dynamic imports](https://basarat.gitbook.io/typescript/main-1/defaultisbad#dynamic-imports): Dynamic imports is when you import modules dynamically during execution instead of at the start using `await import(..)`. 
    Default import: 
    ```javascript
    const HighCharts = await import('https://code.highcharts.com/js/es-modules/masters/highcharts.src.js');
    HighCharts.default.chart('container', { ... }); // Notice `.default`
    ```
    Instead of named import:
    ```javascript
    const {HighCharts} = await import('https://code.highcharts.com/js/es-modules/masters/highcharts.src.js');
    HighCharts.chart('container', { ... }); // Notice `.default`
    ```
6. Linking  [stackOverFlow post](https://stackoverflow.com/questions/33305954/typescript-export-vs-default-export#:~:text=For%20example%2C%20if%20you%20default,in%20all%20the%20other%20files) from the `pros` category because the user is against default exports.
7. In [this article: 4 Best Practices to Write Quality JavaScript Modules](https://dmitripavlutin.com/javascript-modules-best-practices/) from **February 26, 2020**: [1.Prefer named exports](https://dmitripavlutin.com/javascript-modules-best-practices/#1-prefer-named-exports)
8. Complication when creating types file in typescript #118 

## Vote 

What do you prefer:
- **Named exports** 👍 
-  **Default exports** 🎉

**Comment by bb** [comment]
Thanks for the detailed overview! It was not my intention to trigger that much work with my seemingly small question. Sorry.

To elaborate my vote in favor of named exports, some additional comments:

* Regarding *Pro-default 2*

  I think this argument is a bit historic as default was the earlier option. I think this point should be removed from the list or even be moved to the con-default list. Here's why:

  > where you can change the name

  You can also rename named exports like e.g. ```import { instantMeiliSearch as whatIWant } from 'instant-meilisearch'```. Refactoring would then rename instantMeiliSearch but not the local name whatIWant.

  Also, given the package is already installed, with Named exports I could just type `instantM` (or even less), get the autocompletion (from vscode, in my case) which then automatically adds a named import (or extends an existing one, if something else from the package was already imported).

  If I'm missing the import but have a snippet pasted from somewhere like basic usage example from the readme, it would underline the method name and provide the autofix to add a named import. I *think* it does not do that for default exports (at least it didn't for `axios` which I just tried it with.)

  Concluding this bullet, I'd consider named exports easier to use than default and would just cross that point from the pro-list.

* Another con-default:

  If there's a reason later to export another function, this would be very easy with named exports, but not with default exports.


* Another pro-default:

  * changing it now would be breaking change (but the earlier, the better?)



Here's another list of arguments against default export (some of which have already been mentioned), from @basarat's book TypeScript Deep Dive: https://basarat.gitbook.io/typescript/main-1/defaultisbad

**Comment by bidoubiwa** [comment]
Closed by: #238
